### DISCUSSION TITLE
DTO (POCO) type to GraphQL for use in generation of Strawberry Shake mutations

### DISCUSSION THREAD

**Discussion by jkears** [ChilliCream/hotchocolate#4715] [Q&A] [answered]
I would like to know if there exists a convenient helper class within either the Hot Chocolate or Strawberry Shake framework, that I can pass in a DTO (POCO) type, and have it return its GraphQL schema that I can use to code-generate Mutations parameter variables.  

For example, let's assume I have an API service operation named **CreateDailyFloatSheet** which takes in a single parameter of type **DailyFloatSheet** , and where **DailyFloatSheet** is some custom DTO (POCO) type. 

I would like to call the helper class that converts **DailyFloatSheet** into it's equivalent GraphQL type definition.    

The illustration below shows the CreateDailyFloatSheet  API operation accepts a DailyFloatSheet input parameter type ...

![image](https://user-images.githubusercontent.com/1290937/151952515-53a6e890-1932-449e-a108-34015fd6e0f0.png)

The code below would be representative of the server-side mutation operation to be called by Hot Chocolate, along with the type definitions for it's **DailyFloatSheet**  type, input parameter...

```
public async Task<CreateDailyFloatSheetPayload> CreateDailyFloatSheet(DailyFloatSheet dailyFloatSheet)
{
    // Omitted code for brevity
}
 
public class DailyFloatSheet
{
    public string Comment { get; set; }
    public Guid DriverId { get; set; }
    public DateTime Date { get; set; }
    public int Number { get; set; }
    public IEnumerable<EquipmentLocationRecord> EquipmentLocationRecords { get; set; }
    public Guid Id { get; set; }
}

public class EquipmentLocationRecord
{
    public Guid PickupLocationId { get; set; }
    public Guid DropoffLocationId { get; set; }
    public DateTime PickupTimeStamp { get; set; }
    public DateTime DropOffTimeStamp { get; set; }
    public decimal FuelLevel { get; set; }
    public string Comment { get; set; }
    public Guid EquipmentId { get; set; }
    public IEnumerable<Attachment> Attachments { get; set; }
    public IEnumerable<MiscAttachment> MiscAttachments { get; set; }
    public Guid Id { get; set; }
}

public class Attachment
{
    public Guid EquipmentAttachmentId { get; set; }
    public Guid Id { get; set; }
}

public class MiscAttachment
{
    public Guid EquipmentAttachmentId { get; set; }
    public Guid Id { get; set; }
}
```

Ideally, I would utilize the helper class to produce the equivalent GraphQL so that I could code-generate all of the mutations for each API method on an aggregate,  of which, would then be imported by Strawberry Shake and generated into the resultant Strawberry Shake client.   

Any assistance would be greatly appreciated.  Please let me know if you require further information.

**Comment by jkears** [comment, marked as answer]
I have resolved this.

**Reply by goodstas** [reply]
@jkears How did u resolve it?

**Reply by jkears** [reply]
@goodstas, we previously built our own **Domain Modeling tool** that we use to code-generate our microservices (images shared below) and we generated a backing Meta Service in which we inject each domain model's schema.  

This is the modeling tool running within VS code showing our Meta Services Domain model ...
![image](https://user-images.githubusercontent.com/1290937/219414437-758b93dc-956f-4af6-811f-e448d1d267d6.png)

This is a closer shot of the DomainElement aggregate from the Meta Services Model...
![image](https://user-images.githubusercontent.com/1290937/219414952-ca7e5388-f8c6-4473-a2ef-2a0cd896f7db.png)

Then using GraphQL, we query the schema for any Service within any model to generate the above solution.        

![image](https://user-images.githubusercontent.com/1290937/219417014-383230d8-ab5d-4158-b1e2-6557f3d8064c.png)


The above GraphQL query is a meta reflection of this model I am working on for our front-end UI Designer to build Blazor apps ...
![image](https://user-images.githubusercontent.com/1290937/219417466-f26ac39f-7525-4797-9d34-cadf4c98e257.png)

Example query on EditServices WorkSpace Aggregate meta...

![image](https://user-images.githubusercontent.com/1290937/219418643-e92ccec8-039c-4aa1-bc1b-ac718984c92d.png)

The same should be doable using GraphQL introspection, however no one from ChiliCream team provided us guidance on how to do that, so we chose to go with our existing Meta approach, because we have it and can query and filter through any domain model.

**Reply by goodstas** [reply]
Thanks. Looks a little bit complex.
